rasa/core/training/story_conflict.py

```diff
@@ -0,0 +1,235 @@
+            action: Name of the action.
+            story_name: Name of the story where this action is chosen.
+        """
+        self._conflicting_actions[action] += [story_name]
+
+    @property
+    def conflicting_actions(self) -> List[Text]:
+        """List of conflicting actions.
+
+        Returns:
+            List of conflicting actions.
+
+        """
+        return list(self._conflicting_actions.keys())
+
+    @property
+    def has_prior_events(self) -> bool:
+        """Checks if prior events exist.
+
+        Returns:
+            True if anything has happened before this conflict, otherwise False.
+        """
+        return _get_previous_event(self.sliced_states[-1])[0] is not None
+
+    def __str__(self) -> Text:
+        # Describe where the conflict occurs in the stories
+        last_event_type, last_event_name = _get_previous_event(self.sliced_states[-1])
+        if last_event_type:
+            conflict_message = (
+                f"CONFLICT after {last_event_type} '{last_event_name}':\n"
+            )
+        else:
+            conflict_message = f"CONFLICT at the beginning of stories:\n"
+
+        # List which stories are in conflict with one another
+        for action, stories in self._conflicting_actions.items():
+            conflict_message += "  " + self._summarize_conflict(action, stories)
+
+        return conflict_message
+
+    @staticmethod
+    def _summarize_conflict(action, stories):
+        if len(stories) > 3:
+            # Four or more stories are present
+            conflict_description = (
+                f"'{stories[0]}' and {len(stories) - 1} other trackers"
+            )
+        else:
+            conflict_description = (
+                {1: "'{}'", 2: "'{}' and '{}'", 3: "'{}', '{}', and '{}'",}
+                .get(len(stories))
+                .format(*stories)
+            )
+
+        return f"{action} predicted in {conflict_description}\n"
+
+
+class TrackerEventStateTuple(NamedTuple):
+    """Holds a tracker, an event, and sliced states associated with those."""
+
+    tracker: TrackerWithCachedStates
+    event: Event
+    sliced_states: List[Dict[Text, float]]
+
+    @property
+    def sliced_states_hash(self):
+        return hash(str(list(self.sliced_states)))
+
+
+def find_story_conflicts(
+    trackers: List[TrackerWithCachedStates], domain: Domain, max_history: int
+) -> List[StoryConflict]:
+    """Generates a list of StoryConflict objects, describing conflicts in the given trackers.
+
+    Args:
+        trackers: Trackers in which to search for conflicts.
+        domain: The domain.
+        max_history: The maximum history length to be taken into account.
+    Returns:
+        List of conflicts.
+    """
+
+    # We do this in two steps, to reduce memory consumption:
+
+    # Create a 'state -> list of actions' dict, where the state is
+    # represented by its hash
+    state_action_mapping = _find_conflicting_states(trackers, domain, max_history)
```

Comment:
optional: One could thing that these are all states and not just the conflicting ones.
```suggestion
    conflicting_state_action_mapping = _find_conflicting_states(trackers, domain, max_history)
```